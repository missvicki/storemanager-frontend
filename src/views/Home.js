import React, { Component } from 'react';
import { connect, } from "react-redux";

import Products from '../components/Products';
import Navigation from '../components/Navigation';
import {productsFetch} from '../actions/productsAction';
import {usersCreate} from '../actions/createUser';
import CreateUsers from '../components/CreateUsers';
import { removeUserError } from "../actions/index";

export class HomeView extends Component {
    constructor(props) {
        super(props);
        this.state = {
            user_name: "",
            name: "",
            role: "",
            password: ""
        };
    }
    componentDidMount = () => {
        const url =
            "https://store-manager-ap1.herokuapp.com/api/v2/products";
        const payload = {
            url
        };
        this.props.products(payload);
    }

    handleLogout = (event) => {
        event.preventDefault();
        const token = localStorage.getItem("token");
        if (token) {
            if (confirm("Are you sure you want to logout?")) {
                localStorage.clear()
                this.props.history.push("/");
            }
        }
    }

    onSubmitHandler = e => {
        e.preventDefault();
        const { name, user_name, password, role } = this.state;
        const data = { name, user_name, password, role  };
        const url =
            "https://store-manager-ap1.herokuapp.com/api/v2/auth/signup";
        const payload = {
            data,
            history: this.props.history,
            url
        };
        this.props.usersCreate(payload); 
    };

    onDismissHandler = () => {
        const {
            dismissUserError
        } = this.props
        dismissUserError();
    };

    onChangeHandler = e => this.setState({ [e.target.name]: e.target.value });

    render() {
        const productProps = {
            product: this.props.product
        }
        const createProps = {
            name: this.state.name,
            user_name: this.state.product_name,
            role: this.state.category,
            password: this.state.unit_price,
            onSubmitHandler: this.onSubmitHandler,
            onChangeHandler: this.onChangeHandler,
            onDismissHandler: this.onDismissHandler,
            loading: this.props.loading,
            userError: this.props.userError,
            userMessage: this.props.userMessage
        };

        return (
            <div>
                <Navigation onLogoutClick={this.handleLogout} />
                <CreateUsers {...createProps} />
                <Products {...productProps} />
            </div>
        );
    }
}
export const mapDispatchToProps = dispatch => {
    return {
        products: payload => dispatch(productsFetch(payload)),
        usersCreate: payload => dispatch(usersCreate(payload)),
        dismissUserError: () => dispatch(removeUserError())
    };
};

export const mapStateToProps = state => ({
    product: state.products.product,
    userError: state.users.userError,
    loading: state.users.loading,
    userMessage: state.users.userMessage
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(HomeView);
