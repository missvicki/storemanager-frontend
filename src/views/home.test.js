import React from "react";
import { shallow } from "enzyme";
import { HomeView, mapDispatchToProps, mapStateToProps } from "./Home";

describe('Home View', () => {
    const mockProductsFetch = jest.fn();
    const mockRemoveUserError = jest.fn();
    const mockUsersCreate = jest.fn();
    const dispatch = jest.fn();

    const productProps = {
        products: mockProductsFetch,
        dismissUserError: mockRemoveUserError,
        usersCreate: mockUsersCreate,
        loading: false,
        userError: "",
        userMessage: ""

    };

    const homeView = shallow(<HomeView {...productProps} />);
    const homeViewInstance = homeView.instance();

    it("renders homeView component correctly", () => {
        expect(homeView).toMatchSnapshot();
    });

    it("component did mount dispatches productFetch action creator when called", () => {
        homeViewInstance.componentDidMount();
        expect(mockProductsFetch).toHaveBeenCalled();
    });

    it("should handleLogout", () => {
        homeViewInstance.handleLogout({ preventDefault: () => null });
        const token = "jjsjjas";
        expect(token).toBe("jjsjjas");
        if (token) {
            window.confirm = jest.fn();
            window.confirm("are you sure")
            expect(window.confirm).toHaveBeenCalled();
        }
    });

    it("onChangeHandler updates name in state", () => {
        homeViewInstance.onChangeHandler({
            target: { value: "victoria", name: "name" }
        });
        expect(homeView.state().name).toBe("victoria");
    });

    it("onChangeHandler updates user name in state", () => {
        homeViewInstance.onChangeHandler({
            target: { value: "victoria", name: "user_name" }
        });
        expect(homeView.state().user_name).toBe("victoria");
    });

    it("onChangeHandler updates role in state", () => {
        homeViewInstance.onChangeHandler({
            target: { value: "admin", name: "role" }
        });
        expect(homeView.state().role).toBe("admin");
    });

    it("onChangeHandler updates password in state when an event with value of password is passed", () => {
        homeViewInstance.onChangeHandler({
            target: { value: "okay", name: "password" }
        });
        expect(homeView.state().password).toBe("okay");
    });

    it("onDismissHandler dispatches removeUserError action creator to remove login error message", () => {
        homeViewInstance.onDismissHandler();
        expect(mockRemoveUserError).toHaveBeenCalled();
    });

    it("onSubmitHandler dispatches loginFetch action creator when called", () => {
        homeViewInstance.onSubmitHandler({ preventDefault: () => null });
        expect(mockUsersCreate).toHaveBeenCalled();
    });
    it("should map dispatch to props", () => {
        mapDispatchToProps(dispatch).products();
        mapDispatchToProps(dispatch).usersCreate();
        mapDispatchToProps(dispatch).dismissUserError();
        expect(dispatch.mock.calls.length).toBe(3);
    });
});
