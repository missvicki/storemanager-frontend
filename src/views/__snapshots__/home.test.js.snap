// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home View renders homeView component correctly 1`] = `
<div>
  <Navigation
    onLogoutClick={[Function]}
  />
  <CreateUsers
    handleChange={[Function]}
    loading={false}
    name=""
    onChangeHandler={[Function]}
    onDismissHandler={[Function]}
    onSubmitHandler={[Function]}
    password=""
    userError=""
    userMessage=""
    user_name=""
  />
  <Products />
</div>
`;
